PDML Timeout not Respected

When using PDML for long running tasks (> 1 hour) the task fails after one hour with a deadline exceeded exception (`Exception in thread "main" com.google.cloud.spanner.SpannerException: DEADLINE_EXCEEDED: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 3599.998701666s.`). 

Note that by default the timeout should be 2 hours:
https://github.com/googleapis/java-spanner/blob/45acd8960c961d48e91a7b1546efa64d9e9ae576/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java#L313 

This is however not respected and overriden by the RPC timeout configured, which by default is 1 hour:

https://github.com/googleapis/googleapis/blob/50ae1c72fd94a3ae4269394b09e4b7fbb9251146/google/spanner/v1/spanner_grpc_service_config.json#L14

#### Steps to reproduce

  1. Create a default spanner client.
  2. Run a PDML update task that takes more than 1 hour. I have created a table with > 1 billion records to reproduce the problem.

#### Code example

```java
final SpannerOptions options = SpannerOptions.newBuilder().build();
final Spanner spanner = options.getService();
final DatabaseId id = DatabaseId.of("project", "instance", "database");
final DatabaseClient client = spanner.getDatabaseClient(id);

// This times out after 1 hour with a DeadlineExceededException
client.executePartitionedUpdate(Statement.of(
    "UPDATE Table SET Column = 'Value' WHERE true"
));
```

#### Stack trace
```
Exception in thread "main" com.google.cloud.spanner.SpannerException: DEADLINE_EXCEEDED: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 3599.998701666s. [buffered_nanos=115604067, remote_addr=spanner.googleapis.com/142.250.67.10:443]
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted(SpannerExceptionFactory.java:211)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:57)
	at com.google.cloud.spanner.SpannerExceptionFactory.fromApiException(SpannerExceptionFactory.java:226)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:144)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:94)
	at com.google.cloud.spanner.PartitionedDMLTransaction.executeStreamingPartitionedUpdate(PartitionedDMLTransaction.java:156)
	at com.google.cloud.spanner.SessionImpl.executePartitionedUpdate(SessionImpl.java:118)
	at com.google.cloud.spanner.SessionPool$PooledSession.executePartitionedUpdate(SessionPool.java:1349)
	at com.google.cloud.spanner.SessionPool$PooledSessionFuture.executePartitionedUpdate(SessionPool.java:1236)
	at com.google.cloud.spanner.DatabaseClientImpl$3.apply(DatabaseClientImpl.java:227)
	at com.google.cloud.spanner.DatabaseClientImpl$3.apply(DatabaseClientImpl.java:224)
	at com.google.cloud.spanner.DatabaseClientImpl.runWithSessionRetry(DatabaseClientImpl.java:241)
	at com.google.cloud.spanner.DatabaseClientImpl.executePartitionedUpdate(DatabaseClientImpl.java:222)
	at com.google.cloud.sample.app.pdml.dao.UserDao.activateUsers(UserDao.java:61)
	at com.google.cloud.sample.app.pdml.PdmlApp.main(PdmlApp.java:17)
Caused by: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 3599.998701666s. [buffered_nanos=115604067, remote_addr=spanner.googleapis.com/142.250.67.10:443]
	at io.grpc.Status.asRuntimeException(Status.java:533)
	at com.google.api.gax.grpc.GrpcDirectStreamController$ResponseObserverAdapter.onClose(GrpcDirectStreamController.java:149)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.cloud.spanner.spi.v1.SpannerErrorInterceptor$1$1.onClose(SpannerErrorInterceptor.java:100)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
	at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66)
	at io.grpc.internal.ClientCallImpl$1CloseInContext.runInContext(ClientCallImpl.java:416)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

#### Any additional information below
It seems that the following is happening:

- By default we set the `streamWaitTimeout` to 2 hours in the context for the gRPC call.
- By default we set the `streamIdleTimeout` to 30 mins in the context for the gRPC call.
- By default we set the `timeout` to null in the context for the gRPC call.
- During the call to the gRPC endpoint we merge the read context with a default read context, which changes the value of the `timeout` to 1 hour.
- The PDML change fails after 1 hour (not 2 hours) with a deadline exceeded exception, according to the `timeout` variable set.

I am not sure where the `streamWaitTimeout` and `streamIdleTimeout` values are used.

I think that we would need to set the also the `timeout` value based on the `partitionedDmlTimeout` for PDML executions.

--- a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/PartitionedDMLTransaction.java
+++ b/google-cloud-spanner/src/main/java/com/google/cloud/spanner/PartitionedDMLTransaction.java
@@ -78,12 +78,11 @@ class PartitionedDMLTransaction implements SessionTransaction {
    * statement, and will retry the stream if an {@link UnavailableException} is thrown, using the
    * last seen resume token if the server returns any.
    */
-  long executeStreamingPartitionedUpdate(final Statement statement, Duration timeout) {
+  long executeStreamingPartitionedUpdate(final Statement statement, final Duration timeout) {
     checkState(isValid, "Partitioned DML has been invalidated by a new operation on the session");
     log.log(Level.FINER, "Starting PartitionedUpdate statement");
     boolean foundStats = false;
     long updateCount = 0L;
-    Duration remainingTimeout = timeout;
     Stopwatch stopWatch = createStopwatchStarted();
     try {
       // Loop to catch AbortedExceptions.
@@ -106,8 +105,8 @@ class PartitionedDMLTransaction implements SessionTransaction {
             }
           }
           while (true) {
-            remainingTimeout =
-                remainingTimeout.minus(stopWatch.elapsed(TimeUnit.MILLISECONDS), ChronoUnit.MILLIS);
+            Duration remainingTimeout =
+                timeout.minus(stopWatch.elapsed(TimeUnit.MILLISECONDS), ChronoUnit.MILLIS);
             if (remainingTimeout.isNegative() || remainingTimeout.isZero()) {
               // The total deadline has been exceeded while retrying.
               throw new DeadlineExceededException(

--- a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/PartitionedDMLTransaction.java
+++ b/google-cloud-spanner/src/main/java/com/google/cloud/spanner/PartitionedDMLTransaction.java
@@ -19,6 +19,7 @@ package com.google.cloud.spanner;
 import static com.google.common.base.Preconditions.checkState;
 
 import com.google.api.gax.grpc.GrpcStatusCode;
+import com.google.api.gax.rpc.DeadlineExceededException;
 import com.google.api.gax.rpc.ServerStream;
 import com.google.api.gax.rpc.UnavailableException;
 import com.google.cloud.spanner.SessionImpl.SessionTransaction;
@@ -83,7 +84,7 @@ class PartitionedDMLTransaction implements SessionTransaction {
     boolean foundStats = false;
     long updateCount = 0L;
     Duration remainingTimeout = timeout;
-    Stopwatch stopWatch = Stopwatch.createStarted();
+    Stopwatch stopWatch = createStopwatchStarted();
     try {
       // Loop to catch AbortedExceptions.
       while (true) {
@@ -107,6 +108,11 @@ class PartitionedDMLTransaction implements SessionTransaction {
           while (true) {
             remainingTimeout =
                 remainingTimeout.minus(stopWatch.elapsed(TimeUnit.MILLISECONDS), ChronoUnit.MILLIS);
+            if (remainingTimeout.isNegative() || remainingTimeout.isZero()) {
+              // The total deadline has been exceeded while retrying.
+              throw new DeadlineExceededException(
+                  null, GrpcStatusCode.of(Code.DEADLINE_EXCEEDED), false);
+            }
             try {
               builder.setResumeToken(resumeToken);
               ServerStream<PartialResultSet> stream =
@@ -155,6 +161,10 @@ class PartitionedDMLTransaction implements SessionTransaction {
     } catch (Exception e) {
       throw SpannerExceptionFactory.newSpannerException(e);
     }
+  }
+
+  Stopwatch createStopwatchStarted() {
+    return Stopwatch.createStarted();
   }
 
   @Override

--- a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java
+++ b/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java
@@ -309,7 +309,7 @@ public class GapicSpannerRpc implements SpannerRpc {
 
                 // Set a keepalive time of 120 seconds to help long running
                 // commit GRPC calls succeed
-                .setKeepAliveTime(Duration.ofSeconds(GRPC_KEEPALIVE_SECONDS * 1000))
+                .setKeepAliveTime(Duration.ofSeconds(GRPC_KEEPALIVE_SECONDS))
 
                 // Then check if SpannerOptions provides an InterceptorProvider. Create a default
                 // SpannerInterceptorProvider if none is provided
@@ -365,6 +365,7 @@ public class GapicSpannerRpc implements SpannerRpc {
           .setStreamWatchdogProvider(watchdogProvider)
           .executeSqlSettings()
           .setRetrySettings(partitionedDmlRetrySettings);
+      pdmlSettings.executeStreamingSqlSettings().setRetrySettings(partitionedDmlRetrySettings);
       // The stream watchdog will by default only check for a timeout every 10 seconds, so if the
       // timeout is less than 10 seconds, it would be ignored for the first 10 seconds unless we
       // also change the StreamWatchdogCheckInterval.

--- a/google-cloud-spanner/src/test/java/com/google/cloud/spanner/PartitionedDmlTransactionTest.java
+++ b/google-cloud-spanner/src/test/java/com/google/cloud/spanner/PartitionedDmlTransactionTest.java
@@ -1,0 +1,247 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.fail;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyMap;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.google.api.gax.grpc.GrpcStatusCode;
+import com.google.api.gax.rpc.AbortedException;
+import com.google.api.gax.rpc.ServerStream;
+import com.google.api.gax.rpc.UnavailableException;
+import com.google.cloud.spanner.spi.v1.SpannerRpc;
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Ticker;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.ByteString;
+import com.google.spanner.v1.BeginTransactionRequest;
+import com.google.spanner.v1.ExecuteSqlRequest;
+import com.google.spanner.v1.ExecuteSqlRequest.QueryMode;
+import com.google.spanner.v1.PartialResultSet;
+import com.google.spanner.v1.ResultSetStats;
+import com.google.spanner.v1.Transaction;
+import com.google.spanner.v1.TransactionSelector;
+import io.grpc.Status.Code;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.concurrent.TimeUnit;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+import org.threeten.bp.Duration;
+
+@SuppressWarnings("unchecked")
+@RunWith(JUnit4.class)
+public class PartitionedDmlTransactionTest {
+
+  @Mock private SpannerRpc rpc;
+
+  @Mock private SessionImpl session;
+
+  private final String sessionId = "projects/p/instances/i/databases/d/sessions/s";
+  private final ByteString txId = ByteString.copyFromUtf8("tx");
+  private final ByteString resumeToken = ByteString.copyFromUtf8("resume");
+  private final String sql = "UPDATE FOO SET BAR=1 WHERE TRUE";
+  private final ExecuteSqlRequest executeRequestWithoutResumeToken =
+      ExecuteSqlRequest.newBuilder()
+          .setQueryMode(QueryMode.NORMAL)
+          .setSession(sessionId)
+          .setSql(sql)
+          .setTransaction(TransactionSelector.newBuilder().setId(txId))
+          .build();
+  private final ExecuteSqlRequest executeRequestWithResumeToken =
+      executeRequestWithoutResumeToken.toBuilder().setResumeToken(resumeToken).build();
+
+  @Before
+  public void setup() {
+    MockitoAnnotations.initMocks(this);
+    when(session.getName()).thenReturn(sessionId);
+    when(session.getOptions()).thenReturn(Collections.EMPTY_MAP);
+    when(rpc.beginTransaction(any(BeginTransactionRequest.class), anyMap()))
+        .thenReturn(Transaction.newBuilder().setId(txId).build());
+  }
+
+  @Test
+  public void testExecuteStreamingPartitionedUpdate() {
+    ResultSetStats stats = ResultSetStats.newBuilder().setRowCountLowerBound(1000L).build();
+    PartialResultSet p1 = PartialResultSet.newBuilder().setResumeToken(resumeToken).build();
+    PartialResultSet p2 = PartialResultSet.newBuilder().setStats(stats).build();
+    ServerStream<PartialResultSet> stream = mock(ServerStream.class);
+    when(stream.iterator()).thenReturn(ImmutableList.of(p1, p2).iterator());
+    when(rpc.executeStreamingPartitionedDml(
+            Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class)))
+        .thenReturn(stream);
+
+    PartitionedDMLTransaction tx = new PartitionedDMLTransaction(session, rpc);
+    long count = tx.executeStreamingPartitionedUpdate(Statement.of(sql), Duration.ofMinutes(10));
+    assertThat(count).isEqualTo(1000L);
+    verify(rpc).beginTransaction(any(BeginTransactionRequest.class), anyMap());
+    verify(rpc)
+        .executeStreamingPartitionedDml(
+            Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class));
+  }
+
+  @Test
+  public void testExecuteStreamingPartitionedUpdateAborted() {
+    ResultSetStats stats = ResultSetStats.newBuilder().setRowCountLowerBound(1000L).build();
+    PartialResultSet p1 = PartialResultSet.newBuilder().setResumeToken(resumeToken).build();
+    PartialResultSet p2 = PartialResultSet.newBuilder().setStats(stats).build();
+    ServerStream<PartialResultSet> stream1 = mock(ServerStream.class);
+    Iterator<PartialResultSet> iterator = mock(Iterator.class);
+    when(iterator.hasNext()).thenReturn(true, true, false);
+    when(iterator.next())
+        .thenReturn(p1)
+        .thenThrow(
+            new AbortedException(
+                "transaction aborted", null, GrpcStatusCode.of(Code.ABORTED), true));
+    when(stream1.iterator()).thenReturn(iterator);
+    ServerStream<PartialResultSet> stream2 = mock(ServerStream.class);
+    when(stream2.iterator()).thenReturn(ImmutableList.of(p1, p2).iterator());
+    when(rpc.executeStreamingPartitionedDml(
+            any(ExecuteSqlRequest.class), anyMap(), any(Duration.class)))
+        .thenReturn(stream1, stream2);
+
+    PartitionedDMLTransaction tx = new PartitionedDMLTransaction(session, rpc);
+    long count = tx.executeStreamingPartitionedUpdate(Statement.of(sql), Duration.ofMinutes(10));
+    assertThat(count).isEqualTo(1000L);
+    verify(rpc, times(2)).beginTransaction(any(BeginTransactionRequest.class), anyMap());
+    verify(rpc, times(2))
+        .executeStreamingPartitionedDml(
+            Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class));
+  }
+
+  @Test
+  public void testExecuteStreamingPartitionedUpdateUnavailable() {
+    ResultSetStats stats = ResultSetStats.newBuilder().setRowCountLowerBound(1000L).build();
+    PartialResultSet p1 = PartialResultSet.newBuilder().setResumeToken(resumeToken).build();
+    PartialResultSet p2 = PartialResultSet.newBuilder().setStats(stats).build();
+    ServerStream<PartialResultSet> stream1 = mock(ServerStream.class);
+    Iterator<PartialResultSet> iterator = mock(Iterator.class);
+    when(iterator.hasNext()).thenReturn(true, true, false);
+    when(iterator.next())
+        .thenReturn(p1)
+        .thenThrow(
+            new UnavailableException(
+                "temporary unavailable", null, GrpcStatusCode.of(Code.UNAVAILABLE), true));
+    when(stream1.iterator()).thenReturn(iterator);
+    ServerStream<PartialResultSet> stream2 = mock(ServerStream.class);
+    when(stream2.iterator()).thenReturn(ImmutableList.of(p1, p2).iterator());
+    when(rpc.executeStreamingPartitionedDml(
+            Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class)))
+        .thenReturn(stream1);
+    when(rpc.executeStreamingPartitionedDml(
+            Mockito.eq(executeRequestWithResumeToken), anyMap(), any(Duration.class)))
+        .thenReturn(stream2);
+
+    PartitionedDMLTransaction tx = new PartitionedDMLTransaction(session, rpc);
+    long count = tx.executeStreamingPartitionedUpdate(Statement.of(sql), Duration.ofMinutes(10));
+    assertThat(count).isEqualTo(1000L);
+    verify(rpc).beginTransaction(any(BeginTransactionRequest.class), anyMap());
+    verify(rpc)
+        .executeStreamingPartitionedDml(
+            Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class));
+    verify(rpc)
+        .executeStreamingPartitionedDml(
+            Mockito.eq(executeRequestWithResumeToken), anyMap(), any(Duration.class));
+  }
+
+  @Test
+  public void testExecuteStreamingPartitionedUpdateUnavailableAndThenDeadlineExceeded() {
+    PartialResultSet p1 = PartialResultSet.newBuilder().setResumeToken(resumeToken).build();
+    ServerStream<PartialResultSet> stream1 = mock(ServerStream.class);
+    Iterator<PartialResultSet> iterator = mock(Iterator.class);
+    when(iterator.hasNext()).thenReturn(true, true, false);
+    when(iterator.next())
+        .thenReturn(p1)
+        .thenThrow(
+            new UnavailableException(
+                "temporary unavailable", null, GrpcStatusCode.of(Code.UNAVAILABLE), true));
+    when(stream1.iterator()).thenReturn(iterator);
+    when(rpc.executeStreamingPartitionedDml(
+            Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class)))
+        .thenReturn(stream1);
+
+    PartitionedDMLTransaction tx =
+        new PartitionedDMLTransaction(session, rpc) {
+          @Override
+          Stopwatch createStopwatchStarted() {
+            Ticker ticker = mock(Ticker.class);
+            when(ticker.read())
+                .thenReturn(0L, 1L, TimeUnit.NANOSECONDS.convert(10L, TimeUnit.MINUTES));
+            return Stopwatch.createStarted(ticker);
+          }
+        };
+    try {
+      tx.executeStreamingPartitionedUpdate(Statement.of(sql), Duration.ofMinutes(10));
+      fail("missing expected DEADLINE_EXCEEDED exception");
+    } catch (SpannerException e) {
+      assertThat(e.getErrorCode()).isEqualTo(ErrorCode.DEADLINE_EXCEEDED);
+      verify(rpc).beginTransaction(any(BeginTransactionRequest.class), anyMap());
+      verify(rpc)
+          .executeStreamingPartitionedDml(
+              Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class));
+    }
+  }
+
+  @Test
+  public void testExecuteStreamingPartitionedUpdateAbortedAndThenDeadlineExceeded() {
+    PartialResultSet p1 = PartialResultSet.newBuilder().setResumeToken(resumeToken).build();
+    ServerStream<PartialResultSet> stream1 = mock(ServerStream.class);
+    Iterator<PartialResultSet> iterator = mock(Iterator.class);
+    when(iterator.hasNext()).thenReturn(true, true, false);
+    when(iterator.next())
+        .thenReturn(p1)
+        .thenThrow(
+            new AbortedException(
+                "transaction aborted", null, GrpcStatusCode.of(Code.ABORTED), true));
+    when(stream1.iterator()).thenReturn(iterator);
+    when(rpc.executeStreamingPartitionedDml(
+            Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class)))
+        .thenReturn(stream1);
+
+    PartitionedDMLTransaction tx =
+        new PartitionedDMLTransaction(session, rpc) {
+          @Override
+          Stopwatch createStopwatchStarted() {
+            Ticker ticker = mock(Ticker.class);
+            when(ticker.read())
+                .thenReturn(0L, 1L, TimeUnit.NANOSECONDS.convert(10L, TimeUnit.MINUTES));
+            return Stopwatch.createStarted(ticker);
+          }
+        };
+    try {
+      tx.executeStreamingPartitionedUpdate(Statement.of(sql), Duration.ofMinutes(10));
+      fail("missing expected DEADLINE_EXCEEDED exception");
+    } catch (SpannerException e) {
+      assertThat(e.getErrorCode()).isEqualTo(ErrorCode.DEADLINE_EXCEEDED);
+      verify(rpc, times(2)).beginTransaction(any(BeginTransactionRequest.class), anyMap());
+      verify(rpc)
+          .executeStreamingPartitionedDml(
+              Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class));
+    }
+  }
+}

--- a/google-cloud-spanner/src/test/java/com/google/cloud/spanner/PartitionedDmlTransactionTest.java
+++ b/google-cloud-spanner/src/test/java/com/google/cloud/spanner/PartitionedDmlTransactionTest.java
@@ -52,6 +52,8 @@ import org.junit.runners.JUnit4;
 import org.mockito.Mock;
 import org.mockito.Mockito;
 import org.mockito.MockitoAnnotations;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
 import org.threeten.bp.Duration;
 
 @SuppressWarnings("unchecked")
@@ -240,6 +242,55 @@ public class PartitionedDmlTransactionTest {
       assertThat(e.getErrorCode()).isEqualTo(ErrorCode.DEADLINE_EXCEEDED);
       verify(rpc, times(2)).beginTransaction(any(BeginTransactionRequest.class), anyMap());
       verify(rpc)
+          .executeStreamingPartitionedDml(
+              Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class));
+    }
+  }
+
+  @Test
+  public void testExecuteStreamingPartitionedUpdateMultipleAbortsUntilDeadlineExceeded() {
+    PartialResultSet p1 = PartialResultSet.newBuilder().setResumeToken(resumeToken).build();
+    ServerStream<PartialResultSet> stream1 = mock(ServerStream.class);
+    Iterator<PartialResultSet> iterator = mock(Iterator.class);
+    when(iterator.hasNext()).thenReturn(true);
+    when(iterator.next())
+        .thenReturn(p1)
+        .thenThrow(
+            new AbortedException(
+                "transaction aborted", null, GrpcStatusCode.of(Code.ABORTED), true));
+    when(stream1.iterator()).thenReturn(iterator);
+    when(rpc.executeStreamingPartitionedDml(
+            Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class)))
+        .thenReturn(stream1);
+
+    PartitionedDMLTransaction tx =
+        new PartitionedDMLTransaction(session, rpc) {
+          long ticks = 0L;
+
+          @Override
+          Stopwatch createStopwatchStarted() {
+            Ticker ticker = mock(Ticker.class);
+            when(ticker.read())
+                .thenAnswer(
+                    new Answer<Long>() {
+                      @Override
+                      public Long answer(InvocationOnMock invocation) throws Throwable {
+                        return TimeUnit.NANOSECONDS.convert(++ticks, TimeUnit.MINUTES);
+                      }
+                    });
+            return Stopwatch.createStarted(ticker);
+          }
+        };
+    try {
+      tx.executeStreamingPartitionedUpdate(Statement.of(sql), Duration.ofMinutes(10));
+      fail("missing expected DEADLINE_EXCEEDED exception");
+    } catch (SpannerException e) {
+      assertThat(e.getErrorCode()).isEqualTo(ErrorCode.DEADLINE_EXCEEDED);
+      // It should start a transaction exactly 10 times (10 ticks == 10 minutes).
+      verify(rpc, times(10)).beginTransaction(any(BeginTransactionRequest.class), anyMap());
+      // The last transaction should timeout before it starts the actual statement execution, which
+      // means that the execute method is only executed 9 times.
+      verify(rpc, times(9))
           .executeStreamingPartitionedDml(
               Mockito.eq(executeRequestWithoutResumeToken), anyMap(), any(Duration.class));
     }

